### PR TITLE
Add option to pause synchronization on metered connections

### PR DESCRIPTION
**Pull request by erikjv**

### CODE REVIEW

**Comment by erikjv** [comment]
To test on Windows:

Make sure the "Pause..." checkbox is checked:

Screenshot 2023-11-02 at 18 22 43

Check that the account is connected:

Screenshot 2023-11-02 at 18 23 33

Go to the windows network adapter settings by opening the "Settings" app, then choose "Network & Internet", and click the "Properties" button. This should have a switch to turn the connection to metered:

Screenshot 2023-11-02 at 18 25 37

Toggling the switch will change the status of the client to indicate it's paused:

Screenshot 2023-11-02 at 18 26 22

Toggling the switch back should go to the connected state.

File: src/gui/accountsettings.cpp [hunk 1]
```diff
@@ -687,20 +688,43 @@ void AccountSettings::slotScheduleCurrentFolderForceFullDiscovery()
...
+    };
+
     if (auto selectedFolder = this->selectedFolder()) {
-        // Terminate and reschedule any running sync
-        for (auto *folder : FolderMan::instance()->folders()) {
-            if (folder->isSyncRunning()) {
-                folder->slotTerminateSync();
-                FolderMan::instance()->scheduler()->enqueueFolder(folder);
-            }
+        if (isMetered() && ConfigFile().pauseSyncWhenMetered()) {
```
Comment (TheOneRing) [inline src/gui/accountsettings.cpp:699]: At the moment the menu item is disabled to I can't force a sync?

File: src/gui/accountstate.cpp [hunk 2]
```diff
@@ -136,8 +137,18 @@ AccountState::AccountState(AccountPtr account)
                 break;
             }
         });
-    } else {
-        qCWarning(lcAccountState) << "Failed to load QNetworkInformation";
+
+        connect(qNetInfo, &QNetworkInformation::isMeteredChanged, this, [this](bool isMetered) {
+            if (ConfigFile().pauseSyncWhenMetered()) {
+                if (state() == State::Connected && isMetered) {
+                    qCInfo(lcAccountState) << "Network switched to a metered connection, setting account state to PausedDueToMetered";
```
Comment (TheOneRing) [inline src/gui/accountstate.cpp:144]: I guess we need a new icon, currently it looks like we are logged out.
Maybe we should display the paused icon for now?
image
image

File: src/gui/networksettings.cpp [hunk 3]
```diff
@@ -177,6 +181,21 @@ void NetworkSettings::loadBWLimitSettings()
     _ui->uploadSpinBox->setValue(cfgFile.uploadLimit());
 }
 
+void NetworkSettings::loadMeteredSettings()
+{
+    if (QNetworkInformation *qNetInfo = QNetworkInformation::instance()) {
```
Comment (TheOneRing) [inline src/gui/networksettings.cpp:186]: The change only takes effect on the next connection change/client restart.
Should we improve that?

File: src/gui/accountsettings.cpp [hunk 4]
```diff
@@ -687,20 +688,43 @@ void AccountSettings::slotScheduleCurrentFolderForceFullDiscovery()
 
 void AccountSettings::slotForceSyncCurrentFolder()
 {
+    auto isMetered = []() {
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/accountsettings.cpp:691]: done

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (erikjv) [reply src/gui/accountsettings.cpp]: fixed

File: src/gui/networksettings.cpp (hunk 3 above)
Comment (erikjv) [reply src/gui/networksettings.cpp:186]: Why? Other settings are also loaded once.

**Review by TheOneRing — CHANGES_REQUESTED** [review]
Please add a changelog entry

File: src/gui/folder.cpp [hunk 5]
```diff
@@ -806,7 +806,8 @@ void Folder::setVirtualFilesEnabled(bool enabled)
         };
         if (isSyncRunning()) {
             connect(this, &Folder::syncFinished, this, finalizeVfsSwitch, Qt::SingleShotConnection);
-            slotTerminateSync();
+            QString reason;
```
Comment (TheOneRing) [inline src/gui/folder.cpp:809]: Unused var

File: src/gui/accountsettings.cpp [hunk 6]
```diff
@@ -642,19 +643,35 @@ void AccountSettings::slotScheduleCurrentFolderForceFullDiscovery()
...
+{
+    // Terminate and reschedule any running sync
+    for (auto *folder : FolderMan::instance()->folders()) {
+        if (folder->isSyncRunning()) {
+            folder->slotTerminateSync(tr("User triggered force sync"));
+            FolderMan::instance()->scheduler()->enqueueFolder(folder);
+        }
     }
+
+    selectedFolder->slotWipeErrorBlacklist(); // issue #6757
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/accountsettings.cpp:671]: But we add it with high priority, wouldn't that put it in front of the queue?

File: src/gui/folder.h [hunk 7]
```diff
@@ -277,6 +277,9 @@ class Folder : public QObject
     {
         return *_engine;
     }
+
+    void abort(const QString &reason);
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/folder.h:281]: removed

File: src/gui/folder.cpp (hunk 5 above)
Comment (erikjv) [reply src/gui/folder.cpp:809]: removed

File: src/gui/accountsettings.cpp [hunk 8]
```diff
@@ -674,12 +691,19 @@ void AccountSettings::slotAccountStateChanged()
                                .arg(Utility::escape(safeUrl.toString()));
 
     switch (state) {
+    case AccountState::PausedDueToMetered:
+        [[fallthrough]];
     case AccountState::Connected: {
```
[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/accountsettings.cpp:696]: done

File: src/gui/accountsettings.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheOneRing) [reply src/gui/accountsettings.cpp:671]: Yes, but upon terminating the current sync, we might directly start a new sync.
Therefore, we should make sure that the important one is scheduled already.
Of course we should also ensure that we don't kill the sync we just scheduled.

[the remark this answers is not in the document]
Comment (erikjv) [reply src/gui/accountsettings.cpp:674]: Done. Please check if the solution is a good one.

[the remark this answers is not in the document]
Comment (TheOneRing) [reply src/gui/accountsettings.cpp:674]: Looks good, in theory.
Let's check reality.
LGTM

**Review by TheOneRing — DISMISSED** [review]

**Comment by TheOneRing** [comment]
Please fix the changelog:
> file changelog/unreleased/4808: primary issue ID not found

**Review by TheOneRing — APPROVED** [review]

**Comment by saw-jan** [comment]
Tested: https://github.com/owncloud/client/issues/4808#issuecomment-1880431087

**Pull request merged**